contracts/randdrop/src/contract.rs

```diff
@@ -0,0 +1,843 @@
+    let response = match msg {
+        QueryMsg::IsWinner { address } => to_binary(&query_is_winner(deps, address)?)?,
+        QueryMsg::Config {} => to_binary(&query_config(deps)?)?,
+        QueryMsg::MerkleRoot {} => to_binary(&query_merkle_root(deps)?)?,
+        QueryMsg::IsClaimed { address } => to_binary(&query_is_claimed(deps, address)?)?,
+        QueryMsg::RanddropResults {} => to_binary(&query_results(deps)?)?,
+    };
+    Ok(response)
+}
+
+#[allow(clippy::too_many_arguments)]
+fn execute_update_config(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    manager: Option<String>,
+    randdrop_denom: Option<String>,
+    nois_proxy_denom: Option<String>,
+    nois_proxy_amount: Option<Uint128>,
+    nois_proxy_address: Option<String>,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    // check the calling address is the authorised multisig
+    ensure_eq!(info.sender, config.manager, ContractError::Unauthorized);
+
+    let manager = match manager {
+        Some(ma) => deps.api.addr_validate(&ma)?,
+        None => config.manager,
+    };
+    let randdrop_denom = randdrop_denom.unwrap_or(config.randdrop_denom);
+    let nois_proxy_denom = nois_proxy_denom.unwrap_or(config.nois_proxy.price.denom);
+    let nois_proxy_amount = nois_proxy_amount.unwrap_or(config.nois_proxy.price.amount);
+    let nois_proxy_address = match nois_proxy_address {
+        Some(np) => deps.api.addr_validate(&np)?,
+        None => config.nois_proxy.address,
+    };
+
+    let nois_proxy = NoisProxy {
+        price: Coin {
+            denom: nois_proxy_denom,
+            amount: nois_proxy_amount,
+        },
+        address: nois_proxy_address,
+    };
+
+    CONFIG.save(
+        deps.storage,
+        &Config {
+            manager,
+            randdrop_denom,
+            nois_proxy,
+        },
+    )?;
+
+    Ok(Response::new().add_attribute("action", "update_config"))
+}
+
+// This function will call the proxy and ask for the randomness round
+pub fn execute_randdrop(
+    deps: DepsMut,
+    info: MessageInfo,
+    amount: Uint128,
+    proof: Vec<HexBinary>,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    if !MERKLE_ROOT.exists(deps.storage) {
+        return Err(ContractError::MerkleRootAbsent);
+    }
+
+    if PARTICIPANTS.has(deps.storage, &info.sender) {
+        return Err(ContractError::RandomnessAlreadyRequested);
+    }
+
+    if CLAIMED.has(deps.storage, &info.sender) {
+        return Err(ContractError::Claimed {});
+    }
+
+    // The contract will spend NOIS tokens on this info.sender to buy randomness.
+    // To prevent spam and users abusing the funds, we only allow addresses listed on the randdrop to call this entrypoint
+    // Sending the proof here will make sure that the sender is a potential randdrop winner
+    let merkle_root = MERKLE_ROOT.load(deps.storage)?; // This can be optimised by not loading the state twice
```

Comment:
```suggestion
    let merkle_root = MERKLE_ROOT.load(deps.storage)?;
```

using the new `exists` above instead of the old `may_load` makes this unnecessary to worry about